test/mpi/test_mpi_cuda.c

```diff
@@ -0,0 +1,668 @@
+    map_desc.location = pool_props.location;
+    CUDA_CALL(cuMemPoolSetAccess(mpool, &map_desc, 1));
+    CUDA_CALL(cuMemAllocFromPoolAsync(&ptr, size, mpool, NULL));
+    CUDA_CALL(cuStreamSynchronize(NULL));
+
+    alloc_mem.ptr = ptr;
+    alloc_mem.obj = (void*)mpool;
+
+    return alloc_mem;
+}
+
+void free_mempool(alloc_mem_t *alloc_mem)
+{
+    CUDA_CALL(cuMemFree(alloc_mem->ptr));
+    CUDA_CALL(cuMemPoolDestroy((CUmemoryPool)alloc_mem->obj));
+}
+#endif
+
+allocator_t allocators[] = {{"cuMemAlloc", alloc_mem_alloc, free_mem_alloc},
+                            {"cuMemAllocManaged", alloc_mem_alloc_managed,
+                             free_mem_alloc},
+                            {"VMM", alloc_vmm, free_vmm},
+#if HAVE_CUDA_FABRIC
+                            {"mempool", alloc_mempool, free_mempool},
+                            {"VMM_Fabric", alloc_vmm_fabric, free_vmm}
+#endif
+};
+
+void cuda_memcpy(void *dest, const void *src, size_t n)
+{
+    CUDA_CALL(cuMemcpy((CUdeviceptr)dest, (CUdeviceptr)src, n));
+    CUDA_CALL(cuCtxSynchronize());
+}
+
+static int check_result(CUdeviceptr dptr, const void *gold_data, size_t size)
+{
+    void *h_buffer;
+    int ret;
+
+    h_buffer = malloc(size);
+    if (h_buffer == NULL) {
+        fprintf(stderr, "failed to allocate host memory\n");
+        return -1;
+    }
+
+    cuda_memcpy(h_buffer, (void*)dptr, size);
+    ret = memcmp(h_buffer, gold_data, size);
+    free(h_buffer);
+    return ret;
+}
+
+static int test_alloc_prim_send_prim(const allocator_t *allocator_send,
+                                     const allocator_t *allocator_recv,
+                                     const void *gold_data, size_t size,
+                                     CUdevice cu_dev)
+{
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    int res;
+
+    retain_and_push_primary_context(cu_dev);
+
+    alloc_mem_send = allocator_send->alloc(size);
+    alloc_mem_recv = allocator_recv->alloc(size);
+
+    cuda_memcpy((void*)alloc_mem_send.ptr, gold_data, size);
+
+    mpi_pingpong(alloc_mem_send.ptr, alloc_mem_recv.ptr, size);
+
+    res = check_result(alloc_mem_recv.ptr, gold_data, size);
+
+    allocator_send->free(&alloc_mem_send);
+    allocator_recv->free(&alloc_mem_recv);
+
+    pop_and_release_primary_context(cu_dev);
+
+    return res;
+}
+
+static int test_alloc_prim_send_no(const allocator_t *allocator_send,
+                                   const allocator_t *allocator_recv,
+                                   const void *gold_data, size_t size,
+                                   CUdevice cu_dev)
+{
+    alloc_mem_t alloc_mem_send, alloc_mem_recv;
+    CUcontext primary_ctx;
+    int res;
+
+    retain_and_push_primary_context(cu_dev);
+
+    alloc_mem_send = allocator_send->alloc(size);
+    alloc_mem_recv = allocator_recv->alloc(size);
+
+    cuda_memcpy((void*)alloc_mem_send.ptr, gold_data, size);
+
+    CUDA_CALL(cuCtxPopCurrent(&primary_ctx));
```

Comment:
can we add a check that there is no other context set currently?

Reply:
Added.